### ISSUE TITLE
Add code comments to help contributor onboarding

### ISSUE THREAD

**Issue by gabejohnson** [sweet-js/sweet.js#677]
I'm opening this issue as a reminder more than a one-time task.

It took me > 6 months to get familiar with the code base and I'm still weak in areas. Now this may say more about me than it does about the project, but I suspect the latter.

Adding header comments outlining the purpose of a file/class and how it's used by its dependents could go a long way toward easing onboarding for new contributors. Describing how the pieces fit into the overall algorithm would also be really useful. Adding links to portions of or copying pieces of your thesis @disnet could help too.

**Comment by disnet** [comment]
Good idea! The codebase is definitely complex. I'd be holding off doing too much design documentation for a while since pieces were still in flux but I think most things are stable enough for this to be worth it now.
